lib/Serialization/ModuleFile.h

```diff
@@ -344,31 +345,91 @@ class ModuleFile
+
+  /// Emit a generic deserialization error via \c diagnoseFatal().
+  llvm::Error diagnoseFatal() const {
+    return diagnoseFatal(createFatalError());
   }
-  template <typename T> T fatalIfUnexpected(llvm::Expected<T> expected) const {
-    if (expected)
-      return std::move(expected.get());
-    fatal(expected.takeError());
+
+  /// Emit a fatal error via \c diagnoseFatal() and consume it.
+  void diagnoseAndConsumeFatal(llvm::Error error) const {
+    llvm::consumeError(diagnoseFatal(std::move(error)));
+  }
+
+  /// Emit a generic fatal error via \c diagnoseFatal() and consume it.
+  void diagnoseAndConsumeFatal() const {
+    llvm::consumeError(diagnoseFatal());
   }
 
-  /// Report an unexpected format error that could happen only from a memory-level
-  /// inconsistency. Please prefer passing an error to `fatal(llvm::Error error)` when possible.
-  [[noreturn]] void fatal() const {
-    fatal(llvm::make_error<llvm::StringError>(
-        "Memory corruption or serialization format inconsistency.",
-        llvm::inconvertibleErrorCode()));
+  /// Use this in \p void functions as:
+  ///
+  ///    if (diagnoseAndConsumeFatalIfNotSuccess(...)) return;
+  bool diagnoseAndConsumeFatalIfNotSuccess(llvm::Error error) const {
+    if (!error)
+      return false;
+    llvm::consumeError(diagnoseFatal(std::move(error)));
+    return true;
+  }
+
+  /// Use this in functions that return Expected<T> as:
+  ///
+  ///    if (auto error = diagnoseFatalIfNotSuccess(...)) return error;
+  llvm::Error diagnoseFatalIfNotSuccess(llvm::Error error) const {
+    if (!error)
+      return error;
+    return diagnoseFatal(std::move(error));
   }
 
-  [[noreturn]] void fatal(StringRef msg) const {
-    fatal(llvm::make_error<llvm::StringError>(
-          msg,
-          llvm::inconvertibleErrorCode()));
+  /// Emit and return a string error via \c diagnoseFatal().
+  llvm::Error diagnoseFatal(StringRef msg) const {
+    return diagnoseFatal(llvm::make_error<llvm::StringError>(
+        msg, llvm::inconvertibleErrorCode()));
+  }
+
+  /// Emit and consume a string error via \c diagnoseFatal().
+  void diagnoseAndConsumeFatal(StringRef msg) const {
+    return llvm::consumeError(diagnoseFatal(msg));
+  }
+
+
+  /// Report an unexpected format error that could happen only from a
+  /// memory-level inconsistency. Please prefer passing an error to
+  /// `fatal(llvm::Error error)` when possible.
+  static llvm::Error createFatalError(
+      llvm::StringRef msg =
+          "Memory corruption or serialization format inconsistency.") {
+    return llvm::make_error<llvm::StringError>(msg,
+                                               llvm::inconvertibleErrorCode());
+  }
+
+  /// Emit a fatal error and abort.  This function is deprecated, try to use
+  /// diagnoseFatal() instead. Clients such as LLDB really prefer not to be
+  /// killed.
+//  LLVM_DEPRECATED("Use diagnoseFatal and pass up the error instead.",
+//                  "diagnoseFatal")
```

Comment:
indentation is off

Reply:
oh I see now this isn't a comment, it's a commented out macro call

Reply:
Yeah LLVM_DEPRECATED is too new for the main branch, but also too cumbersome to cherry-pick because the commit changes llvm::Optional.

Reply:
> are you planning to do follow up changes to remove more or all of their uses
Yes.